compiler/hash-intrinsics/src/intrinsics.rs

```diff
@@ -46,52 +96,727 @@ impl Debug for DefinedIntrinsics {
+                            // @@Todo: properly handle f32
+                            let lhs: f64 = env.try_use_term_as_float_lit(lhs).unwrap();
+                            let rhs: f64 = env.try_use_term_as_float_lit(rhs).unwrap();
+                            Ok(env.new_bool_term(operate_float!(parsed_op, lhs, rhs)))
+                        }
+                        LitTy::F64 => {
+                            let lhs: f64 = env.try_use_term_as_float_lit(lhs).unwrap();
+                            let rhs: f64 = env.try_use_term_as_float_lit(rhs).unwrap();
+                            Ok(env.new_bool_term(operate_float!(parsed_op, lhs, rhs)))
+                        }
+                        LitTy::Bool => {
+                            let lhs: bool = env.try_use_term_as_bool(lhs).unwrap();
+                            let rhs: bool = env.try_use_term_as_bool(rhs).unwrap();
+                            Ok(env.new_bool_term(operate_bool!(parsed_op, lhs, rhs)))
+                        }
+                        LitTy::Char => {
+                            let lhs: char = env.try_use_term_as_char_lit(lhs).unwrap();
+                            let rhs: char = env.try_use_term_as_char_lit(rhs).unwrap();
+                            Ok(env.new_bool_term(operate_char!(parsed_op, lhs, rhs)))
+                        }
+                    },
+                    None => Err(INVALID_OP.to_string()),
+                }
+            },
+        )
+    }
+
+    /// Add the `endo_bin_op` intrinsic.
+    ///
+    /// This intrinsic has signature
+    /// `endo_bin_op: (T: Type, op: u8, a: T, b: T) -> T`
+    ///
+    /// The `op` parameter is an integer which represents the operation to
+    /// perform, which is one of the `EndoBinOp` variants. The `a` and `b`
+    /// parameters are the two operands, and the return value is the result
+    /// of the operation.
+    fn add_endo_bin_op_intrinsic<T: AccessToEnv + AccessToPrimitives + Copy>(
+        env: T,
+        implementations: &DefaultPartialStore<IntrinsicId, Intrinsic>,
+    ) -> FnDefId {
+        let t_sym = env.new_symbol("T");
+        let op_sym = env.new_symbol("op");
+        let a_sym = env.new_symbol("a");
+        let b_sym = env.new_symbol("b");
+        let params = env.param_utils().create_params(
+            [
+                ParamData { default: None, name: t_sym, ty: env.new_small_universe_ty() },
+                ParamData {
+                    default: None,
+                    name: op_sym,
+                    ty: env.new_data_ty(env.primitives().u8()),
+                },
+                ParamData { default: None, name: a_sym, ty: env.new_ty(t_sym) },
+                ParamData { default: None, name: b_sym, ty: env.new_ty(t_sym) },
+            ]
+            .into_iter(),
+        );
+        let ret = env.new_ty(t_sym);
+
+        Self::add_intrinsic(
+            env,
+            implementations,
+            "endo_bin_op",
```

Comment:
"endo" meaning what?

Reply:
Just that it has the same return type as its arguments